### ISSUE TITLE
intent to patch: changes to Node.js MSI for Windows

### ISSUE THREAD

**Issue by othiym23** [joyent/node#8528]
I would like to see two small changes made to the Node installer to better support npm users who want to upgrade their local installation of npm:

1. Update `npm.cmd` to have the following behavior:
    1. if `.\node_modules\npm\bin\npm-cli.js` exists, run that
    2. if `%APPDATA%\npm\node_modules\npm\bin\npm-cli.js` exists, run that
    3. otherwise, run the version of npm installed alongside `node.exe`
2. install `npmrc` to `%INSTALLDIR%\etc\npmrc` (and / or `%APPDATA%\etc\npmrc`) instead of `%INSTALLDIR%\node_modules\npm`

This allows users to update their version of npm without Administrator privileges or having to change their system's PATHs, and also leaves the version of npm installed with node there as a fallback if their local install ends up broken.

I (and the rest of the npm team) are happy to do the work necessary to update the MSI ourselves, but would like to know if this patch is likely to be accepted before spending time on this.

See npm/npm#6413 for motivation and related discussion.

**Comment by smikes** [comment]
Does this need to be an issue on `node` (and presumably also `iojs`) ? 

What if `npm` just changes the contents of the distributed `bin/npm.cmd`"?

**Comment by piscisaureus** [comment]
> What about
> > install npmrc to %INSTALLDIR%\etc\npmrc (and / or %APPDATA%\etc\npmrc) instead of %INSTALLDIR%\node_modules\npm
> 
> This does require changes to the MSI. If we could install the Windows-specific configuration to a stable, known location outside the npm package, it would simplify things a lot for corporate IT. Do you have any guidance or concerns around that?

Depends what you want. Placing it in `%appdata%` makes it a per-user setting, which may or may not be desirable. If you put it in `%INSTALLDIR%` it is still tied to the node installation; you might also put it in `%ProgramData%` to share the settings file between users.

The canonical location for installed files is "vendor\\application",  e,g, "c:\\program files\\mozilla\\firefox" or "%appdata%\\microsoft\\office". Because open source software often doesn't really have a vendor, that part is often omitted. But I wouldn't recommend using "%appdata%\\etc" for anything.

**Comment by othiym23** [comment]
My goal was to come up with a stable location to put the per-installation configuration that didn't require any changes to npm. ;) What I *want* is to put the data in a consistent place such that it will persist across global installations of npm. There's some logic built into npm to make it [copy this configuration across upgrades](https://github.com/npm/npm/blob/master/lib/build.js#L64-L79) of npm when that file is in the install directory, and this issue exists in part because that logic was broken for a while. It would be nice to have a standard location outside of the npm install path to put that, because that technique feels brittle to me. It would also be nice to have that location be backwards- and forwards-compatible. `$PREFIX\etc\npmrc` allows us to take advantage of how npm deals with this on Linux and OS X; it's definitely not a solution that's idiomatically natural on Windows.

There has been an issue open on npm for a while (npm/npm#4564) trying to come up with the best place for npm to put its globally-installed stuff. Maybe the answer is to put the global configuration there as well (although that isn't a backwards-compatible solution).

**Comment by orangemocha** [comment]
It seems that the proposed solution is to change the way that npm.cmd locates npm-cli.js. Does any of that require a change to the Node MSI? Are the proposed npm changes blocked on https://github.com/npm/npm/issues/4564?

**Comment by joaocgreis** [comment]
@othiym23 I have a patched version of `npm.cmd` that looks first in `npm prefix -g`. It does not change npmrc but works if the prefix is changed. It should solve the problem with `npm install -g npm`. Should I open a PR in npm (against what branch?) or is this blocked on something else?

**Comment by jasnell** [comment]
@othiym23 ... any reason to keep this one open still?

**Comment by joaocgreis** [comment]
I believe all the changes discussed here already landed on npm, so this can be closed. Please reopen if I missed something.
